processor/vote.go

```diff
@@ -0,0 +1,172 @@
+package processor
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/consensys/gnark/std/algebra/emulated/sw_bn254"
+	"github.com/consensys/gnark/std/math/emulated"
+	gnarkecdsa "github.com/consensys/gnark/std/signature/ecdsa"
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+	"github.com/iden3/go-iden3-crypto/mimc7"
+	"github.com/vocdoni/arbo"
+	"github.com/vocdoni/vocdoni-z-sandbox/circuits"
+	"github.com/vocdoni/vocdoni-z-sandbox/circuits/voteverifier"
+	"github.com/vocdoni/vocdoni-z-sandbox/crypto"
+	"github.com/vocdoni/vocdoni-z-sandbox/log"
+	"github.com/vocdoni/vocdoni-z-sandbox/storage"
+	"github.com/vocdoni/vocdoni-z-sandbox/storage/census"
+	"github.com/vocdoni/vocdoni-z-sandbox/types"
+)
+
+type VoteProcessor struct {
+	stg    *storage.Storage
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewVoteProcessor creates a new VoteProcessor instance with the given storage
+// instance.
+func NewVoteProcessor(stg *storage.Storage) *VoteProcessor {
+	return &VoteProcessor{
+		stg: stg,
+	}
+}
+
+// Start method starts the vote processor. It will process ballots in the
+// background. It iterates over the ballots available in the storage and
+// generates proofs of the validity of the ballots, storing them back in the
+// storage. It will stop processing ballots when the context is cancelled.
+func (p *VoteProcessor) Start(ctx context.Context) error {
+	p.ctx, p.cancel = context.WithCancel(ctx)
+	// run the verifier in background until its context is cancelled
+	go func() {
+		for {
+			select {
+			case <-p.ctx.Done():
+				return
+			default:
+				// get the next ballot
+				ballot, key, err := p.stg.NextBallot()
+				if err != nil {
+					if err != storage.ErrNoMoreElements {
+						log.Errorf("failed to get next ballot: %v", err)
+					}
+					break
+				}
+				log.Debugf("new ballot to process for address %s", ballot.Address.String())
+				// process the ballot
+				verifiedBallot, err := p.ProcessBallot(ballot)
+				if err != nil {
+					log.Errorf("failed to process ballot: %v", err)
+					break
+				}
+				log.Debugf("ballot processed for address %s", ballot.Address.String())
+				// store the verified ballot
+				if err := p.stg.MarkBallotDone(key, verifiedBallot); err != nil {
+					log.Errorf("failed to mark ballot done: %v", err)
+				}
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}()
+	return nil
+}
+
+// Stop method cancels the context of the vote processor, stopping the
+// processing of ballots.
+func (p *VoteProcessor) Stop() error {
+	p.cancel()
+	return nil
+}
+
+func (p *VoteProcessor) ProcessBallot(b *storage.Ballot) (*storage.VerifiedBallot, error) {
+	// decode the process id
+	pid := new(types.ProcessID)
```

Comment:
You can simplify it with:

`p.stg.Process(new(types.ProcessID).SetBytes(b.ProcessID))`